### PR TITLE
Persistent collections updates (part 9)

### PR DESCRIPTION
**Pull request by lorentey**
Add structural `union` and `symmetricDifference` implementations. Add combinatorial tests for all these & fix all issues uncovered. Optimize these operations not to copy any data if they end up returning `self` verbatim.

I think this PR completes all the immediate non-trivial coding work on `PersistentSet` & `PersistentDictionary` -- except for dictionary merge operations; those will be relatively straightforward adaptations of `union`.

I'll still need to add the customary generic overloads for `intersection`/`union`/etc that take arbitrary sequences -- but those are mostly trivial.

These types are shaping up pretty well!

### Checklist
- [X] I've read the [Contribution Guidelines](/README.md#contributing-to-swift-collections)
- [X] My contributions are licensed under the [Swift license](/LICENSE.txt).
- [X] I've followed the coding style of the rest of the project.
- [X] I've added tests covering all new code paths my change adds to the project (if appropriate).
- [ ] I've added benchmarks covering new functionality (if appropriate).
- [X] I've verified that my change does not break any existing tests or introduce unexplained benchmark regressions.
- [ ] I've updated the documentation if necessary.

### CODE REVIEW

**Comment by lorentey** [comment]
I don't have any benchmarks this time around; at least not yet. I'm thinking we could demonstrate the superiority of CHAMPs over flat hash tables by looking at how fast `PersistentSet.subtracting` is when given two sets that are slightly mutated copies of the same original.

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
```
swift-collections/Sources/PersistentCollections/Node/_Bitmap.swift:135:18: warning: instance method '_rank(ofBit:)' cannot be used in an '@inlinable' function because '_CollectionsUtilities' was not imported by this file; this is an error in Swift 6
    _Slot(_value._rank(ofBit: bucket.value))
                 ^
```

Oops

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
Welp

```
error: emit-module command failed due to signal 6 (use -v to see invocation)error: compile command failed due to signal 6 (use -v to see invocation)/Users/ec2-user/jenkins/workspace/pr-swift-collections-macos/branch-release/1.1/swift-collections/Tests/BitCollectionsTests/BitSetTests.swift: error: file failed to round-trip
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
[...]
1.	Apple Swift version 5.8-dev (LLVM ed4d811c2897d77, Swift 8491f52d478aca6)
2.	Compiling with the current language version
3.	While evaluating request ParseSourceFileRequest(source_file "/Users/ec2-user/jenkins/workspace/pr-swift-collections-macos/branch-release/1.1/swift-collections/Tests/BitCollectionsTests/BitSetTests.swift")
4.	With parser at source location: /Users/ec2-user/jenkins/workspace/pr-swift-collections-macos/branch-release/1.1/swift-collections/Tests/BitCollectionsTests/BitSetTests.swift:1278:1
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001085f22b7 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x00000001085f14c5 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x00000001085f2900 SignalHandler(int) + 288
3  libsystem_platform.dylib 0x00007fff20398d7d _sigtramp + 29
4  libdyld.dylib            0x00007fff2036dce8 _dyld_fast_stub_entry(void*, long) + 65
5  libsystem_c.dylib        0x00007fff202a8406 abort + 125
6  swift-frontend           0x0000000104874686 swift::ParseSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 2454
7  swift-frontend           0x0000000104874b41 swift::SimpleRequest<swift::ParseSourceFileRequest, swift::SourceFileParsingResult (swift::SourceFile*), (swift::RequestFlags)12>::evaluateRequest(swift::ParseSourceFileRequest const&, swift::Evaluator&) + 17
8  swift-frontend           0x00000001046f43d7 llvm::Expected<swift::ParseSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::ParseSourceFileRequest>(swift::ParseSourceFileRequest const&) + 471
9  swift-frontend           0x00000001046f3fcd llvm::Expected<swift::ParseSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::ParseSourceFileRequest, (void*)0>(swift::ParseSourceFileRequest const&) + 253
10 swift-frontend           0x00000001046d5cec swift::ParseSourceFileRequest::OutputType swift::evaluateOrDefault<swift::ParseSourceFileRequest>(swift::Evaluator&, swift::ParseSourceFileRequest, swift::ParseSourceFileRequest::OutputType) + 60
11 swift-frontend           0x00000001046cd1c8 swift::SourceFile::getTopLevelDecls() const + 120
12 swift-frontend           0x0000000103fc5ef1 swift::performImportResolution(swift::SourceFile&) + 1169
13 swift-frontend           0x000000010321e9d1 swift::CompilerInstance::performParseAndResolveImportsOnly() + 929
14 swift-frontend           0x000000010321eac3 swift::CompilerInstance::performSema() + 19
15 swift-frontend           0x00000001031b1f38 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3864
16 swift-frontend           0x0000000102fe8d34 swift::mainEntry(int, char const**) + 2932
17 libdyld.dylib            0x00007fff2036ef3d start + 1
18 libdyld.dylib            0x0000000000000033 start + 18446603339975692535
[5/39] Compiling SortedCollectionsTests SortedDictionary Tests.swift
error: fatalError
```

**Comment by lorentey** [comment]
```
Tests/BitCollectionsTests/BitSetTests.swift: error: file failed to round-trip
```

@DougGregor are you collecting these?

**Review by msteindorfer — CHANGES_REQUESTED** [review]
Change List:
* Check / verify potential oversight in structural `union`.

Otherwise, LGTM.

File: Sources/PersistentCollections/Node/_Node+Structural union.swift [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+              copied = true
+            }
+          }
+          else if r.childMap.contains(bucket) {
+            let rslot = r.childMap.slot(of: bucket)
+            let lp = l.itemPtr(at: lslot)
+            let rp = r.childPtr(at: rslot)
+            let h = _Hash(lp.pointee.key)
+            if !rp.pointee.containsKey(level.descend(), lp.pointee.key, h) {
+              node.ensureUniqueAndPushItemIntoNewChild(
```
[the remark this answers is not in the document]
Comment (lorentey) [reply Sources/PersistentCollections/Node/_Node+Structural union.swift]: Excellent catch! I completely missed this in the initial implementation, but it did not escape the power set tests -- 4a4c4a75fd0b0f222fa70615dc3bee5dbf1730b4 fixes it a couple commits later. 

(The other big face palm moment was when I realized I forgot to add any buckets from the right hand node that are missing from the left one. This initial implementation for union was probably my worst first attempt in the whole effort. 🙃)

I'm putting quite a lot of faith in the assumption that the tests are covering all the edge cases. This is stuff is subtle.

**Comment by lorentey** [comment]
It occurs to me we should have in-place mutating variants for `formIntersection`/`formUnion` et al., instead of unconditionally delegating to the copying variants. This would be along the lines of `remove`/`removing` (which could actually use a bit of refactoring, now that we have `Builder` -- which is incidentally also badly in need of an interface cleanup). I'm not sure I want to deal with the pain of duplicating & adapting all this code just yet, though. 

(Like with `remove`, we must not call the `mutating` variants on potentially non-uniquely held nodes to avoid violating the Law of Exclusivity, and unconditionally uniquing nodes along the way would cancel out some/most of the structural benefits.)

**Comment by lorentey** [comment]
@swift-ci test
